File: Leetcode/python/Convert Sorted Array to Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param num, a list of integers
    # @return a tree node
    def sortedArrayToBST(self, num):
        if not num:
            return None
        n = len(num)
        root_val = num[n / 2]
        root = TreeNode(root_val)
        root.left = self.sortedArrayToBST(num[:n / 2])
        root.right = self.sortedArrayToBST(num[n / 2 + 1:])
        return root

File: Leetcode/Algorithm/cpp/Convert Sorted Array to Binary Search Tree.cc
/**
* Definition for binary tree
* struct TreeNode {
* int val;
* TreeNode *left;
* TreeNode *right;
* TreeNode(int x) : val(x), left(NULL), right(NULL) {}
* };
*/
class Solution {
public:
	TreeNode *sortedArrayToBST(vector<int> &num) {
		int n = num.size();
		return sortedArrayToBST(num, 0, n - 1);
	}
private:
	TreeNode *sortedArrayToBST(vector<int> &num, int l, int r) {
		if (l > r) {
			return NULL;
		} else if (l == r) {
			return new TreeNode(num[l]);
		} else {
			int m = (l + r) >> 1;
			TreeNode* root = new TreeNode(num[m]);
			root -> left = sortedArrayToBST(num, l, m - 1);
			root -> right = sortedArrayToBST(num, m + 1, r);
			return root;
		}
	}
};